feat: Introduce sortable Web Analytics tables

rafaeelaudibert opened the pull request:
## Problem

Web Analytics doesn't let you sort your tables very easily.

Closes #28614 

## Changes

We can now sort our web analytics tables. I didn't use the existing sort feature because it relies on tweaking the `query` which isn't something we want here, we need to change the state inside the `webAnalyticsLogic` and then expose that inside the tiles so that it can properly refresh.

Backend implementation is a little finicky but I believe we won't have any problems. Tests cover almost all possible cases and they are passing just fine.

I'm hiding this behind a FF but I intend to enable it 100% from the get-go, this is just a way to easily revert in case UI breaks heavily.

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Manually + unit tests + snapshots

rafaeelaudibert committed:
feat: Introduce sortable Web Analytics tables

We can now sort our web analytics tables. I didn't use the existing sort
feature because it relies on tweaking the `query` which isn't something
we want here, we need to change the state inside the `webAnalyticsLogic`
and then expose that inside the tiles so that it can properly refresh.

Backend implementation is a little finicky but I believe we won't have
any problems. Tests cover almost all possible cases and they are passing
just fine.

I'm hiding this behind a FF but I intend to enable it 100% from the
get-go, this is just a way to easily revert in case UI breaks heavily.

rafaeelaudibert committed:
tests: Add `WEB_ANALYTICS_TABLE_SORTING` FF to snapshots

robbie-c commented:
<img width="1427" alt="Screenshot 2025-02-13 at 13 31 27" src="https://github.com/user-attachments/assets/d73eda0b-680a-499f-8ead-af58fc1ad22d" />

I got this locally

rafaeelaudibert committed:
fix: Fix broken bounce rate

rafaeelaudibert commented on frontend/src/scenes/web-analytics/TableSortingIndicator.tsx:
```diff
@@ -0,0 +1,42 @@
+        <LemonButton
+            type="secondary"
+            size="small"
+            icon={<IconSort className={clsx({ 'rotate-180': tablesOrderBy[1] === 'ASC' })} />}
```
Problem is, the icon is descending by default :)

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tiles/WebAnalyticsTile.stories.tsx:
```diff
@@ -46,6 +52,11 @@ const meta: Meta<typeof Query> = {
 
 const Template: StoryFn<typeof Query> = (args) => {
+    const { setTablesOrderBy } = useActions(webAnalyticsLogic)
+    useEffect(() => {
+        setTablesOrderBy(WebAnalyticsOrderByFields.Views, 'DESC')
+    }, [setTablesOrderBy])
```
Not needed, this is what I intend here, this is for testing purposes

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tiles/WebAnalyticsTile.tsx:
```diff
@@ -330,52 +372,52 @@ export const webAnalyticsDataTableQueryContext: QueryContext = {
         },
         bounce_rate: {
-            title: <span className="pr-5">Bounce Rate</span>,
+            renderTitle: (props) => <span className="pr-5">Bounce Rate{JSON.stringify(props)}</span>,
```
@robbie-c this is why you're seeing that locally, fixed it now

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -429,6 +437,17 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
             },
         ],
+        tablesOrderBy: [
+            null as WebAnalyticsOrderBy | null,
+            persistConfig,
+            {
+                setTablesOrderBy: (_, { orderBy, direction }) => [orderBy, direction],
+                clearTablesOrderBy: () => null,
+
+                // Reset the order by when the conversion goal changes because most of the columns are different
+                setConversionGoal: () => null,
+            },
+        ],
```
This is tricky, might do that in the future

rafaeelaudibert commented on posthog/hogql_queries/web_analytics/stats_table.py:
```diff
@@ -348,6 +364,37 @@ def _main_inner_query(self, breakdown):
 
         return query
 
+    def _order_by(self, columns: list[str]) -> list[ast.OrderExpr] | None:
+        if self.query.orderBy:
+            field, direction = self.query.orderBy
+            column = None
+            if field == WebAnalyticsOrderByFields.VISITORS:
+                column = "context.columns.visitors"
+            elif field == WebAnalyticsOrderByFields.VIEWS:
+                column = "context.columns.views"
+            elif field == WebAnalyticsOrderByFields.CLICKS:
+                column = "context.columns.clicks"
+            elif field == WebAnalyticsOrderByFields.BOUNCE_RATE:
+                column = "context.columns.bounce_rate"
+            elif field == WebAnalyticsOrderByFields.AVERAGE_SCROLL_PERCENTAGE:
+                column = "context.columns.average_scroll_percentage"
+            elif field == WebAnalyticsOrderByFields.SCROLL_GT80_PERCENTAGE:
+                column = "context.columns.scroll_gt80_percentage"
+            elif field == WebAnalyticsOrderByFields.TOTAL_CONVERSIONS:
+                column = "context.columns.total_conversions"
+            elif field == WebAnalyticsOrderByFields.UNIQUE_CONVERSIONS:
+                column = "context.columns.unique_conversions"
+            elif field == WebAnalyticsOrderByFields.CONVERSION_RATE:
+                column = "context.columns.conversion_rate"
+
+            if column is not None and column in columns:
+                # Add a secondary sort by breakdown_value to ensure consistent ordering
+                return [
+                    ast.OrderExpr(expr=ast.Field(chain=[column]), order=direction),
+                    ast.OrderExpr(expr=ast.Field(chain=["context.columns.breakdown_value"]), order="ASC"),
+                ]
+        return None
```
This is intended

rafaeelaudibert commented on posthog/hogql_queries/web_analytics/web_goals.py:
```diff
@@ -223,6 +228,18 @@ def calculate(self):
             )
             results.append([action_name, action_unique, action_total, (current_action_rate, previous_action_rate)])
 
+        if self.query.orderBy is not None:
+            index = None
+            if self.query.orderBy[0] == WebAnalyticsOrderByFields.CONVERTING_USERS:
+                index = 1
+            elif self.query.orderBy[0] == WebAnalyticsOrderByFields.TOTAL_CONVERSIONS:
+                index = 2
+            elif self.query.orderBy[0] == WebAnalyticsOrderByFields.CONVERSION_RATE:
+                index = 3
+
+            if index is not None:
+                results.sort(key=lambda x: x[index], reverse=self.query.orderBy[1] == "DESC")
```
They'll be sorted in the right way because `current` is in front.

rafaeelaudibert commented on posthog/hogql_queries/web_analytics/test/test_web_stats_table.py:
```diff
@@ -1579,3 +1582,238 @@ def test_conversion_rate(self):
         ] == response.columns
+
+    def test_sorting_by_visitors(self):
+        s1 = str(uuid7("2023-12-01"))
+        s2 = str(uuid7("2023-12-01"))
+        s3 = str(uuid7("2023-12-01"))
+
+        self._create_events(
+            [
+                ("p1", [("2023-12-01", s1, "/path1")]),
+                ("p2", [("2023-12-01", s2, "/path1"), ("2023-12-01", s2, "/path2")]),
+                ("p3", [("2023-12-01", s3, "/path1"), ("2023-12-01", s3, "/path2"), ("2023-12-01", s3, "/path3")]),
+            ]
+        )
+
+        # Test ascending order
+        results = self._run_web_stats_table_query(
+            "all",
+            "2023-12-15",
+            breakdown_by=WebStatsBreakdown.PAGE,
+            orderBy=(WebAnalyticsOrderByFields.VISITORS, "ASC"),
+        ).results
+
+        assert [row[0] for row in results] == ["/path3", "/path2", "/path1"]
+
+        # Test descending order
+        results = self._run_web_stats_table_query(
+            "all",
+            "2023-12-15",
+            breakdown_by=WebStatsBreakdown.PAGE,
+            orderBy=(WebAnalyticsOrderByFields.VISITORS, "DESC"),
+        ).results
+
+        assert [row[0] for row in results] == ["/path1", "/path2", "/path3"]
+
```
This is tested in other tests cases

rafaeelaudibert committed:
fix: Small fixes

rafaeelaudibert committed:
fix: Remove dependency

rafaeelaudibert committed:
fix: Python types

rafaeelaudibert commented:
@robbie-c fixed :)

rafaeelaudibert committed:
fix: Use proper cast technique

github-actions[bot] committed:
Update query snapshots

robbie-c commented on posthog/hogql_queries/web_analytics/stats_table.py:
```diff
@@ -233,6 +233,16 @@ def to_path_scroll_bounce_query(self) -> ast.SelectQuery:
+
+        # Compute query order based on the columns we're selecting
+        # Use our default order if we don't have any columns to order by
+        columns = [select.alias for select in query.select if isinstance(select, ast.Alias)]
```
I think it's be good to handle all of the order by logic in one place, so either here or in _order_by

You're already passing columns into _order_by so it should have all the information to do.

robbie-c replied:
Not strongly! I do think that we probably want the sort order to have more to it than [sort, breakdown_value] though.

E.g. I would want
* if I sort by views, it should be [views, visitors, breakdown_value]
* if I sort by visitor, it should be [visitors, views, breakdown_value]

The way I describe seemed to me to be the easiest way to achieve this but there's many other ways too :)

rafaeelaudibert commented on frontend/src/scenes/web-analytics/tiles/WebAnalyticsTile.tsx:
```diff
@@ -268,6 +271,45 @@ const BreakdownValueCell: QueryContextColumnComponent = (props) => {
+        const isAscending = tablesOrderBy?.[1] === 'ASC'
+        const isDescending = tablesOrderBy?.[1] === 'DESC'
+
+        // Toggle between DESC, ASC, and no sort, in this order
```
> I found this UX a bit confusing, especially given we don't have different icons for unset and DESC.

This would be solved if we had better sorting icons, it's actually crazy we don't.

> And if you click on a different header it's > off

I tried that but I thought toggling it off was too hard? I added something to the navbar to make it easier to clear them, so maybe that's enough? I'll try that

> That example also only shows the sort icon when active for that column, or when hovered, which I don't hate, but I don't have strong opinions on :)

That might hurt discoverability 😢. I'll try and make them even more muted so that they're still there, but "less active".

rafaeelaudibert committed:
refactor: Update Web Analytics sort behavior

The behavior is now improved by not having a "disabled" state anymore, you need to clear it on the navbar. The hover design displays on hover only to make it slightly cleaner.

rafaeelaudibert committed:
test: Fix weird bug

For some reason Storybook couldn't find `WebAnalyticsOrderByFields.Views` so let's hardcode it

rafaeelaudibert committed:
refactor: Simplify order by queries

This can be much simpler by checking if columns exist or not. We can see that the previous behavior is kept because snapshots haven't changed.

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

rafaeelaudibert committed:
layout: Align header/row icons

rafaeelaudibert committed:
refactor(types): Fix types for mypy

This is getting out of hand

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Import from correct folder

github-actions[bot] committed:
Update query snapshots
